plugins/NebulaTextures/src/gui/NebulaTexturesDialog.cpp

```diff
@@ -0,0 +1,490 @@
+{
+   QString apiKey = ui->lineEditApiKey->text();
+   QString imagePath = ui->lineEditImagePath->text();
+
+   if (imagePath.isEmpty() || apiKey.isEmpty())
+   {
+      updateStatus("Please provide both API key and image path.");
+      return;
+   }
+
+   // Create the JSON object
+   QJsonObject json;
+   json["apikey"] = apiKey;
+
+   // Convert JSON object to JSON document
+   QJsonDocument jsonDoc(json);
+   QByteArray jsonData = jsonDoc.toJson();
+
+   // Create the POST request URL
+   QUrl url(API_URL+"api/login");
+
+   // Create the form data for x-www-form-urlencoded
+   QUrlQuery query;
+   query.addQueryItem("request-json", QString(jsonData));
+
+   // Set the URL with query
+   url.setQuery(query);
+
+   // Create the POST request
+   QNetworkRequest request(url);
+   request.setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");
+
+   // Send the POST request
+   QNetworkReply *reply = networkManager->post(request, query.toString().toUtf8());
+
+   // Connect the reply finished signal to a slot
+   connect(reply, &QNetworkReply::finished, this, [this, reply]() {onLoginReply(reply);});
+
+   // Perform image upload logic here, e.g., using an API call
+   updateStatus("Sending requests...");
+}
+
+void NebulaTexturesDialog::updateStatus(const QString &status)
+{
+   ui->statusText->setText(status);
+}
+
+
+void NebulaTexturesDialog::onLoginReply(QNetworkReply *reply)
+{
+   QByteArray content = reply->readAll();
+   qDebug()<<content;
+   if (reply->error() != QNetworkReply::NoError) {
+      qDebug() << "Login failed:" << reply->errorString();
+      updateStatus("Login failed!");
+      reply->deleteLater();
+      return;
+   }
+   updateStatus("Login success...");
+
+   QJsonDocument doc = QJsonDocument::fromJson(content);
+   QJsonObject json = doc.object();
+   session = json["session"].toString();
+   qDebug() << "Session ID:" << session;
+
+   QUrl uploadUrl(API_URL + "api/upload");
+   QNetworkRequest request(uploadUrl);
+   QJsonObject uploadJson;
+   uploadJson["session"] = session;
+   uploadJson["allow_commercial_use"] = "d";
+   uploadJson["allow_modifications"] = "d";
+   uploadJson["publicly_visible"] = "y";
+
+   QFile imageFile(ui->lineEditImagePath->text());
+   if (!imageFile.open(QIODevice::ReadOnly)) {
+      qDebug() << "Failed to open image file!";      
+      updateStatus("Failed to open image file!");
+      reply->deleteLater();
+      return;
+   }
+
+   QString boundary_key;
+   // Generate 19 random digits (0-9)
+   for (int i = 0; i < 19; ++i) {
+      boundary_key += QString::number(QRandomGenerator::global()->bounded(10));  // Generates a random digit (0-9)
+   }
+
+   // Format the boundary as '===============<random_number>=='
+   QString boundary1 = "'===============" + boundary_key + "==";
+   QByteArray boundary = boundary1.toUtf8();
+
+
+   QString contentType = QString("multipart/form-data; boundary=%1").arg(boundary);
```

Comment:
What is this conversion to UTF-8 and back to QString supposed to do? Seems like an unnecessary roundtrip.

Reply:
This question is still relevant. The conversion to UTF-8 byte array and then back to `QString` is completely unnecessary. You can form the string as a `QString`, and only when passing it to `setHeader()` do the single conversion.

Reply:
thanks，I got it